File: Ex3_Reductions/dotprod/cpp/yourwork/dotprod_yourwork.cpp
#include <iostream>
#include <vector>

#define ORD 1 << 29 // size of arrays

using namespace std;

int main()
{
    // declare variables
    vector<double> v1, v2; // vectors
    double r;              // this will hold random numbers
    double dotp;           // this will hold the dot product
    int i;                 // index that runs over vectors

    // start random number generator, fill up vectors
    srand(time(NULL));
    for (i = 0; i < ORD; i++)
    {
        r = double(rand() % RAND_MAX) / RAND_MAX; // random number between 0 and 1
        v1.push_back(r);
        r = double(rand() % RAND_MAX) / RAND_MAX;
        v2.push_back(r);
    }

    // calculate dot product
    dotp = 0.0;
    for (i = 0; i < ORD; i++)
    {
        dotp = dotp + v1[i] * v2[i];
    }

    // print result
    cout << "Dot product is: " << dotp << endl;

    // goodbye
    return 0;
}

File: Ex3_Reductions/dotprod/fortran/yourwork/dotprod_yourwork.f90
program dotprod_yourwork
    use, intrinsic :: iso_fortran_env
    implicit none
    integer, parameter :: dp = REAL64   ! double precision
    integer, parameter :: ORD = 2**29   ! arrays size
    real(dp), dimension(:), allocatable :: v1, v2   ! arrays
    real(dp) :: dotp    ! this will hold the dot product
    integer :: i        ! this will run loops over the arrays

    ! allocate arrays
    allocate(v1(ORD))
    allocate(v2(ORD))

    ! attribute values to arrays (random numbers)
    call random_seed()  ! start RNG
    call random_number(v1)
    call random_number(v2)

    ! calculate dot product
    dotp = 0.0_dp
    do i = 1, ORD
        dotp = dotp + v1(i)*v2(i)
    enddo

    ! print result
    write(*,*) 'Dot product is: ', dotp

    ! cleanup and goodbye
    deallocate(v1,v2)

end program dotprod_yourwork